### PR TITLE
Fix index out of bounds when writing FST to different metaOut (#12697)

### LINKED ISSUE #12697: ArrayIndexOutOfBoundsException when writing the FSTStore-backed FST with different DataOutput for meta
### Description

After writing the FSTStore-backed FST to DataOutput, and specifying a different DataOutput for meta, if we try to read from these (using the FST public ctor) we will get the following the exception:

```
java.lang.ArrayIndexOutOfBoundsException: Index 17 out of bounds for length 17

	at __randomizedtesting.SeedInfo.seed([CBCB30F6D2F8FEA1:821F24747AC56DDD]:0)
	at org.apache.lucene.store.ByteArrayDataInput.readVLong(ByteArrayDataInput.java:133)
	at org.apache.lucene.util.fst.FST.<init>(FST.java:494)
	at org.apache.lucene.util.fst.FST.<init>(FST.java:443)
```

The reason is that, when writing to metadata, if the FST is backed by FSTStore, it would not write the numBytes: https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/util/fst/FST.java#L555-L562

The numBytes is instead written by FSTStore to the main DataOutput: https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/util/fst/OnHeapFSTStore.java

Thus if we set the metaOut and dataOut as the same DataOutput, they will subsequently write the numBytes correctly. However if we use different DataOutput, the metaOut will thus lack of the numBytes and cause the index out of bounds exception.

To illustrate:

When writing on the same DataOutput

```
[ HEADER ] [ EMPTY_OUTPUT_FLAG ] [ EMPTY_OUTPUT ] [INPUT_TYPE ] [ START_NODE ] [ NUM_BYTES ] [ MAIN ]
```

When writing on the different DataOutput

```
metaOut: [ HEADER ] [ EMPTY_OUTPUT_FLAG ] [ EMPTY_OUTPUT ] [INPUT_TYPE ] [ START_NODE ]
dataOut: [ NUM_BYTES ] [ MAIN ]
```

Expected DataOutput:
```
metaOut: [ HEADER ] [ EMPTY_OUTPUT_FLAG ] [ EMPTY_OUTPUT ] [INPUT_TYPE ] [ START_NODE ] [ NUM_BYTES ] 
dataOut: [ MAIN ]
```

I can put a fix to this

### Version and environment details

_No response_

### PR DESCRIPTION
**Pull request by dungba88**
### Description

Fix #12697. This will move the writing of numBytes from `OnHeapFSTStore.writeTo()` to `FST.save()`.

`OnHeapFSTStore.size()` will also be modified to return only the numBytes.

As such, this might change the existing behavior. However searching through the Lucene code, `size()` is only referenced for `OffHeapFSTStore`, which already use a correct implementation. I think it could be backported to 9.9

### CODE REVIEW

File: lucene/CHANGES.txt [hunk 1]
```diff
@@ -325,6 +325,8 @@ Bug Fixes
 
 * GITHUB#12571: Fix HNSW graph read bug when built with excessive connections. (Ben Trent).
 
+* GITHUB#12698: Fix index out of bounds bug when saving FSTStore-backed FST with different DataOutput for metadata
```
Comment (mikemccand) [inline lucene/CHANGES.txt:328]: Do we really never, in Lucene today, write the FST and its metadata to different files?

Comment (mikemccand) [reply lucene/CHANGES.txt]: Also (unless you don't want to!), please put your name in the CHANGES entry -- we try to give credit to contributors!

Maybe reword to something like `Fix IndexOutOfBoundsException when saving FSTStore-backed FST with different DataOutput for metadata` or so?

**Review by mikemccand — APPROVED** [review]
Great catch, sneaky bug!  I guess nowhere in Lucene are we using this particular way of saving FST...

I just left small comments about the `CHANGES` entry.

This can go back to 9.x right?

File: lucene/CHANGES.txt (hunk 1 above)
Comment (dungba88) [reply lucene/CHANGES.txt]: > Do we really never, in Lucene today, write the FST and its metadata to different files?

It seems we do, in those `LuceneXXBlockTreeTermsWriter.TermsWriter.finish()`. But they had used the BytesStore-backed FST (FST on-the-fly), so it would not cause the issue either.

I think this issue is to make the two branches more consistent and help to simplify the consolidation of BytesStore and FSTStore.

**Comment by dungba88** [comment]
> This can go back to 9.x right?

I think that too.

I rebased the change, re-added the assertion and updated the CHANGES log :)

**Comment by mikemccand** [comment]
Excellent, thanks @dungba88 -- I'll merge & backport soon.

**Comment by mikemccand** [comment]
Thanks @dungba88 -- I merged to `main` and `branch_9x`!
